### PR TITLE
feat(behavior_velocity): occlusion spot generate not detection area occupancy grid

### PR DESCRIPTION
**Pull request by taikitanaka3**
## Description

- new feature
  - generate not detection area occupancy grid from dynamic object

concept moving vehicle will interrupt obstacle running out behind moving vehicle
Screenshot from 2022-04-01 16-50-21

white: occupied
black: freespace
gray: unkown

image

This PR is going to 
- consider delay of dynamic object briefly
This PR wont 
- consider delay of dynamic object accurately

## Related links

https://github.com/tier4/autoware_launch/pull/282

## Tests performed

by psim

## Notes for reviewers

testing raycast all vehicle

test raycast for all vehicle
Screenshot from 2022-04-01 16-39-49

test raycast only moving vehicle
Screenshot from 2022-04-01 16-50-21

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_velocity_planner/config/occlusion_spot.param.yaml [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
       filter_occupancy_grid: true           # [-] whether to filter occupancy grid by morphologyEx or not
       use_object_info: true                 # [-] whether to reflect object info to occupancy grid map or not
+      use_moving_object_ray_cast: true      # [-] whether to reflect moving object ray_cast to occupancy grid map or not
       use_partition_lanelet: true           # [-] whether to use partition lanelet map data
       pedestrian_vel: 1.5                   # [m/s] assume pedestrian is dashing from occlusion at this velocity
       pedestrian_radius: 0.3                # [m] assume pedestrian width(0.2m) + margin(0.1m)
       debug:                                # !Note: default should be false for performance
         is_show_occlusion: false            # [-] whether to show occlusion point markers.
-        is_show_cv_window: false            # [-] whether to show open_cv debug window
+        is_show_cv_window: true            # [-] whether to show open_cv debug window
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/config/occlusion_spot.param.yaml:14]: will reset this change before merge

Comment (taikitanaka3) [inline planning/behavior_velocity_planner/config/occlusion_spot.param.yaml:14]: will reset this change before merge

**Review by yukkysaito — DISMISSED** [review]

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/grid_utils.cpp [hunk 2]
```diff
@@ -247,6 +236,198 @@ void getCornerPositions(
...
 }
+
+boost::optional<Polygon2d> generateOcclusionPolygon(
+  const Polygon2d & occupancy_poly, const Point2d & origin, const Point2d & min_theta_pos,
+  const Point2d & max_theta_pos, const double ray_max_length = 100.0)
+{
+  using tier4_autoware_utils::normalizeRadian;
+  const double origin_x = origin.x();
+  const double origin_y = origin.y();
+  const double delay_angle = M_PI / 6.0;
```
Comment (0x126) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/grid_utils.cpp:247]: What is "delay_angle" ? Is it ok to fix it at 30°?

Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/grid_utils.cpp:247]: @0x126 
It's not concretely defined yet this is the simplest way of avoiding dynamic object occlusion detection.
To define delay below is necessary .
- distance to object
- relative velocity to dynamic object
- relative angle
- estimated perception delay 
But with this info change of struct grid_utils should be changed and not yet tested.

so I remain this for TODO task in another PR.

Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/grid_utils.cpp:248]: [5988206](https://github.com/autowarefoundation/autoware.universe/pull/620/commits/5988206b0a4e6f514fb7e1775960199816aec632)

@0x126 
I added comment here

**Review by 0x126 — APPROVED** [review]

**Pull request merged**